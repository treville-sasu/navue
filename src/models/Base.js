export class Model {
  // eslint-disable-next-line no-unused-vars
  constructor({ type, ...properties } = {}) {
    Object.assign(this, properties);
  }

  toJSON() {
    return {
      ...this,
      type: this.constructor.name
    };
  }

  // TODO write specs
  clone() {
    return this.constructor.from(JSON.parse(JSON.stringify(this)));
  }

  static from({ type, ...properties } = {}) {
    if (arguments[0] instanceof this) return arguments[0];
    else if (type == this.name) return new this(properties);
    else throw `Invalid data : 'type' should be '${this.name}' got '${type}'`;
  }
}

// TODO : implementation for when Vuejs Suppots Array-like objects.
// export class Store extends Array {
export class Store {
  constructor(properties = {}, ...items) {
    this.items = items;
    Object.assign(this, properties);
  }

  // get items() {
  //   return [...this];
  // }

  [Symbol.iterator]() {
    return this.items[Symbol.iterator]();
  }

  get length() {
    return this.items.length;
  }

  valueOf() {
    return this.items;
  }

  get properties() {
    // eslint-disable-next-line no-unused-vars
    let { items, ...props } = { ...this };
    return props;
    // return Object.fromEntries(
    //   Object.entries(this).filter(([key]) => isNaN(key))
    // );
  }

  get first() {
    return this.items[0];
  }

  get last() {
    return this.items[this.length - 1];
  }

  add(item, index) {
    index ? this.items.splice(index, 0, item) : this.items.push(item);
    return this.length;
  }

  remove(item, index) {
    if (index === undefined) index = this.items.indexOf(item);
    return index >= 0 ? this.items.splice(index, 1).length == 1 : false;
  }

  keep(filter) {
    let i = this.items.length;
    while (i--) {
      let result = filter.call(this.items, this.items[i]);

      if (result == undefined) this.items.splice(i, 1);
      else if (!Object.is(result, this.items[i]))
        this.items.splice(i, 1, result);
    }
    return this;
  }

  toJSON() {
    return {
      ...this.properties,
      items: this.items,
      type: this.constructor.name
    };
  }

  static from({ type, items = [], ...properties }, ...constructors) {
    if (type != this.name)
      throw `Invalid data : 'type' should be '${this.name}' got '${type}'`;
    else {
      let constructor = constructors.shift();
      return new this(
        properties,
        ...items.map(i => {
          return constructor ? constructor.from(i, ...constructors) : i;
        })
      );
    }
  }
}
