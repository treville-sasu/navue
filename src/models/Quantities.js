import { Model } from "@/models/Base.js";

export class Quantity extends Model {
  // eslint-disable-next-line no-unused-vars
  constructor(value, _unit, { type, unit, precision, ...properties } = {}) {
    if (arguments[0] instanceof Quantity) return arguments[0];
    else {
      super(properties);

      if (_unit) this.unit = _unit;
      else this._unit = undefined;

      if (value || value === 0) this.value = value;
      else {
        this._value = undefined;
        this.precision = undefined;
      }

      if (precision) this.precision = precision;

      if (unit) this.unit = unit;
    }
  }

  get value() {
    return this.constructor.significantValue(
      this._to(this._value),
      this.precision
    );
  }

  set value(val) {
    let f = Number(val);
    if (Number.isNaN(f)) throw `\`${val}\` could not be coerced to a number`;
    else {
      this._value = this._from(f);
      this.precision = Math.max(3, this.constructor.significantFigures(val));
    }
  }

  get unit() {
    return this._unit || this.constructor.baseUnit;
  }

  set unit(val) {
    if (!this.constructor.units[val]) throw `'${val}' is not an available unit`;
    else this._unit = val;
  }

  get factor() {
    return this.constructor.units[this.unit];
  }

  _from(val, unit = this.unit) {
    return Number(val) / this._factor(unit);
  }

  _to(val, unit = this.unit) {
    return Number(val) * this._factor(unit);
  }

  _factor(unit) {
    if (this.constructor.units[unit]) return this.constructor.units[unit];
    else throw `'${unit}' is not an available unit`;
  }

  to(unit) {
    return this._to(this.value, unit);
  }

  as(unit, precision) {
    return Object.assign(this.constructor.from(this), { unit, precision });
  }

  add(other, props) {
    if (typeof other == "number" || other instanceof this.constructor)
      return new this.constructor(this + other, undefined, {
        precision: Math.min(this.precision, other.precision),
        unit: this._unit,
        ...props
      });
    else throw "operands should be of the same type";
  }

  // FIXME : prod with quantity should not return the type from this nor other.
  prod(other, props) {
    return new this.constructor(this * other, undefined, {
      precision: Math.min(
        this.precision,
        other.precision || this.constructor.significantFigures(other)
      ),
      unit: this._unit,
      ...props
    });
  }

  valueOf() {
    return this._value;
  }

  toString() {
    let sf = Math.max(1, Math.min(this.precision, 21));
    return `${
      isNaN(this.value)
        ? "-"
        : this.value.toLocaleString(undefined, {
            minimumSignificantDigits: sf,
            maximumSignificantDigits: sf
          })
    } ${this.unit}`;
  }

  toJSON() {
    // eslint-disable-next-line no-unused-vars
    let { _value, _unit, ...properties } = this;
    return {
      ...properties,
      value: this.value,
      unit: this.unit,
      type: this.constructor.name
    };
  }

  static significantFigures(n) {
    let s;
    if (n === 0) {
      return 0;
    } else if (typeof n === "number") {
      s = Math.abs(n).toExponential();
    } else if (typeof n === "string" && !Number.isNaN(Number(n))) {
      s = n.replace(/^\+|-/, "").replace(/^0+/, "");
    } else return;

    s = s.replace(".", "");
    const i = s.lastIndexOf("e");
    return i > 0 ? i : s.length;
  }

  static significantValue(val, precision = 1) {
    if (val === 0) return 0;
    const fact = Math.pow(
      10,
      precision - Math.floor(Math.log(val < 0 ? -val : val) / Math.LN10) - 1
    );
    return Math.round(val * fact) / fact;
  }

  static get units() {
    throw "list of available units should be set on constructor.";
  }

  static get baseUnit() {
    return Object.keys(this.units)[Object.values(this.units).indexOf(1)];
  }

  static from({ type, value, unit, ...properties } = {}) {
    if (!type || this.name == type) return new this(value, unit, properties);
  }
}

export class Distance extends Quantity {
  static get units() {
    return {
      m: 1,
      Km: 0.001,
      ft: 3.28084,
      NM: 0.000539957,
      mi: 0.000621371,
      yd: 1.09361
    };
  }
}

export class Altitude extends Distance {
  constructor(value, unit, { reference, ...properties } = {}) {
    if (unit == "FL") reference = "QNE";
    else if (unit && !reference)
      throw `property 'reference' should be given, got '${reference}'`;

    super(value, unit, { ...properties, reference });
  }

  // densityAltitude(temp) {
  //   return new this.constructor(
  //     this.value + 36.576 * (temp - 15),
  //     "m",
  //     "MSL"
  //   );
  // }

  get value() {
    if (this.unit == "FL")
      return this.constructor.significantValue(this._to(this._value), 2);
    else return this._to(this._value);
  }

  set value(val) {
    super.value = val;
  }

  toString() {
    if (this.unit == "FL") {
      return `FL${String(this.value).padStart(3, "0")}`;
    } else if (this.value && this.reference) {
      return `${super.toString()} ${this.reference}`;
    }
  }

  static get references() {
    return ["MSL", "AGL", "QNH", "QNE", "QFE", "WGS84"];
  }

  static get units() {
    return {
      m: 1,
      ft: 3.28084,
      FL: 0.0328084
    };
  }
}

export class Speed extends Quantity {
  static get units() {
    return {
      "m/s": 1,
      "km/h": 3.6,
      "ft/min": 196.85,
      kt: 1.94384,
      mph: 2.23694
    };
  }

  // TAS(altitude, temperature) {}
}

export class Volume extends Quantity {
  static get units() {
    return {
      m3: 1,
      L: 1000,
      "gal US": 264.172,
      "gal Imp": 219.969
    };
  }
}

export class Consumption extends Volume {
  static get references() {
    return ["h", "1000ft", "u"];
  }

  toString(precision) {
    return `${super.toString(precision)}/${this.reference}`;
  }
}

export class Weight extends Quantity {
  static get units() {
    return {
      kg: 1,
      lb: 2.20462,
      t: 0.001,
      "L-avgas": 1 / 0.721,
      "L-JetA1": 1 / 0.8
    };
  }
}

export class Angle extends Quantity {
  static get units() {
    return {
      rad: 1,
      "°": 180 / Math.PI,
      gon: 200 / Math.PI,
      rev: 1 / 2 / Math.PI,
      "₥": 6400 / 2 / Math.PI
    };
  }
}

export class Azimuth extends Angle {
  get value() {
    //FIXME: 0 should give 360
    return this._to((this._value + 2 * Math.PI) % (2 * Math.PI));
  }

  set value(val) {
    super.value = val;
  }
}

export class Bearing extends Angle {
  get value() {
    //FIXME: 180 % should give 180
    return this._to(((this._value + 3 * Math.PI) % (2 * Math.PI)) - Math.PI);
  }

  set value(val) {
    super.value = val;
  }
}

export class Pressure extends Quantity {
  static get units() {
    return {
      Pa: 1,
      hPa: 0.01,
      Psi: 1 / 6894.76,
      atm: 1 / 101325,
      bar: 0.00001,
      inHg: 1 / 3386.389
    };
  }
}

// TODO Spec it !
export class Temperature extends Quantity {
  static get units() {
    return {
      "°C": [1, 0],
      K: [1, 273.15],
      "°F": [9 / 5.32]
    };
  }

  _from(val, unit = this.unit) {
    return Number(val) / this._factor(unit)[0] + this._factor(unit)[1];
  }

  _to(val, unit = this.unit) {
    return (Number(val) - this._factor(unit)[1]) * this._factor(unit)[0];
  }
}
export class Frequency extends Quantity {
  static get units() {
    return {
      Hz: 1,
      KHz: 0.001,
      MHz: 0.00001,
      GHz: 0.00000001
    };
  }
}
