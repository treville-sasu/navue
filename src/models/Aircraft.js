import { Model, Store } from "@/models/Base.js";
import { Speed, Consumption, Weight, Distance } from "@/models/Quantities.js";

export class Aircraft extends Model {
  constructor({
    registration,
    manufacturer,
    model,
    paces = new Store(),
    consumptions = new Store(),
    balance = new Store({ date: undefined }),
    envelopes = new Store(),
    checklists = new Store()
  } = {}) {
    paces.keep(i => (i instanceof Speed ? i : undefined));
    consumptions.keep(i => (i instanceof Consumption ? i : undefined));
    balance.keep(i => (i instanceof Weight ? i : undefined));
    envelopes.keep(i => (i instanceof Store ? i : undefined));
    // envelopes = envelopes.keep(i => {
    //   return i.keep(j => (j instanceof Weight ? j : undefined));
    // });
    checklists.keep(i => (i instanceof Store ? i : undefined));

    super({
      registration,
      manufacturer,
      model,
      paces,
      consumptions,
      balance,
      envelopes,
      checklists
    });
  }

  static from({
    _id,
    _rev,
    type,
    registration,
    manufacturer,
    model,
    paces,
    consumptions,
    balance,
    envelopes,
    checklists
  } = {}) {
    if (arguments[0] instanceof this) return arguments[0];
    else if (type != this.name)
      throw `Invalid data : 'type' should be '${this.name}' got '${type}'`;
    else {
      if (paces) paces = Store.from(paces, Speed);
      if (consumptions) consumptions = Store.from(consumptions, Consumption);
      if (balance) balance = Store.from(balance, Weight);
      if (envelopes) envelopes = Store.from(envelopes, Store, Moment);
      if (checklists) checklists = Store.from(checklists, Store);

      let imported = new this({
        registration,
        manufacturer,
        model,
        paces,
        consumptions,
        balance,
        envelopes,
        checklists
      });
      if (_id && _rev) Object.assign(imported, { _id, _rev });

      return imported;
    }
  }
}

export class Moment extends Model {
  constructor(mass, lever, properties) {
    if (!mass) mass = new Weight();
    if (!lever) lever = new Distance();

    super({ mass, lever, ...properties });
  }

  get value() {
    return this.mass.value * this.lever.value;
  }

  get _value() {
    return this.mass * this.lever;
  }

  get unit() {
    return `${this.mass.unit}.${this.lever.unit}`;
  }

  valueOf() {
    return this._value;
  }

  toJSON() {
    return {
      ...this,
      value: this.value,
      unit: this.unit,
      type: this.constructor.name
    };
  }
  toCoords() {
    return {
      x: this.lever.value,
      y: this.mass.value
    };
  }

  static from({ type, mass, lever } = {}) {
    if (arguments[0] instanceof this) return arguments[0];
    else if (type == this.name) {
      if (mass) mass = Weight.from(mass);
      if (lever) lever = Distance.from(lever);
      return new this(mass, lever, { type });
    } else throw `Supplied object is no ${this.name}`;
  }
}
