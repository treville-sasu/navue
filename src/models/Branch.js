import { Model } from "@/models/Base.js";
import { Distance } from "@/models/Quantities.js";

import { lineString, featureCollection, round } from "@turf/helpers";
import bbox from "@turf/bbox";
import length from "@turf/length";
import buffer from "@turf/buffer";
import truncate from "@turf/truncate";

export class Branch extends Model {
  constructor(member, properties = {}, ...features) {
    if (member)
      features.forEach((f, i, arr) => {
        arr[i] = member.from(f);
      });
    super({ ...featureCollection(features), properties, member });
  }

  first() {
    return this.features[0];
  }

  last() {
    return this.features[this.features.length - 1];
  }

  push(...positions) {
    return this.features.push(
      ...positions.map(p => (this.member ? this.member.from(p) : p))
    );
  }

  shift() {
    return this.features.shift();
  }

  insert(position, index) {
    const pos = isNaN(index) ? this.features.length : index;
    this.features.splice(pos, 0, position);
    return position;
  }

  remove(index) {
    this.features.splice(index, 1);
    return this.features.length;
  }

  forEach(cb) {
    return this.features.forEach(cb);
  }

  map(cb) {
    return this.features.map(cb);
  }

  filter(cb) {
    return this.features.filter(cb);
  }

  pairs(cb) {
    let collection = [];
    if (this.features.length)
      this.features.reduce((previous, current, i, a) => {
        if (previous) collection.push(cb.call(this, previous, current, i, a));
        return current;
      });
    return collection;
  }

  get geom() {
    return this.map(({ geometry: { coordinates } }) => coordinates);
  }

  get bbox() {
    return bbox(lineString(this.geom));
  }

  get length() {
    return new Distance(
      round(length(lineString(this.geom), { units: "meters" }))
    );
  }

  toJSON() {
    // eslint-disable-next-line no-unused-vars
    let { member, ...json } = super.toJSON();
    return json;
  }

  toGeoJSON(type) {
    switch (type) {
      case "Feature":
      case "LineString":
        return lineString(this.geom, this.properties, { bbox: this.bbox });
      case "MultiPolygon":
        return featureCollection(
          this.pairs((start, end) => {
            return truncate(
              buffer(
                // TODO : Should we create midpoints and average the properties ? let's see in real life
                lineString(
                  [start.geometry.coordinates, end.geometry.coordinates],
                  start.properties
                ),
                Number(start.properties.accuracy) || 15,
                {
                  units: "meters",
                  steps: 1
                }
              ),
              { mutate: true }
            );
          }),
          { bbox: this.bbox }
        );
      case "FeatureCollection":
      default:
        return {
          ...featureCollection(this.features, {
            bbox: this.bbox
          }),
          properties: this.properties
        };
    }
  }

  static from({ type, features = [], properties }, constructor) {
    if (arguments[0] instanceof this) return arguments[0];

    if (type === this.name || type === "FeatureCollection")
      return new this(constructor, properties, ...features);
    else
      throw `Invalid data : 'type' should be '${this.name}' or 'FeatureCollection' got '${type}'`;
  }
}
