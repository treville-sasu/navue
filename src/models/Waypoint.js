import { Model } from "@/models/Base.js";
import { Altitude, Distance, Azimuth } from "@/models/Quantities.js";

import { point, round } from "@turf/helpers";
import rhumbDistance from "@turf/rhumb-distance";
import rhumbBearing from "@turf/rhumb-bearing";
import rhumbDestination from "@turf/rhumb-destination";
import truncate from "@turf/truncate";
import greatCircle from "@turf/great-circle";

export class Waypoint extends Model {
  constructor([longitude, latitude, altitude] = [], properties = {}) {
    const coordinates = [round(longitude, 6), round(latitude, 6)];

    if (altitude !== undefined && !properties["altitude"]) {
      properties["altitude"] = new Altitude(altitude, "m", {
        reference: "WGS84"
      });
    } else if (properties["altitude"]) {
      properties["altitude"] = Altitude.from(properties["altitude"]);
      altitude = round(Number(altitude));
    }

    if (altitude) coordinates.push(round(Number(altitude)));
    super(point(coordinates, properties));
  }

  set longitude(val) {
    this.geometry.coordinates[0] = round(val, 6);
  }

  set latitude(val) {
    this.geometry.coordinates[1] = round(val, 6);
  }

  get altitude() {
    return this.geometry.coordinates[2];
  }

  set altitude(val) {
    this.geometry.coordinates[2] = round(val);
    this.properties.altitude = val;
  }

  get lngLat() {
    return {
      lng: this.geometry.coordinates[0],
      lat: this.geometry.coordinates[1]
    };
  }

  set lngLat({ lng, lat }) {
    this.longitude = lng;
    this.latitude = lat;
  }

  distanceTo(wp) {
    return new Distance(
      round(
        rhumbDistance(this.geometry.coordinates, wp.geometry.coordinates, {
          units: "meters"
        })
      )
    );
  }

  bearingTo(wp) {
    return new Azimuth(
      rhumbBearing(this.geometry.coordinates, wp.geometry.coordinates),
      "°"
    );
  }

  greatCircleTo(wp, properties) {
    let gc = truncate(
      greatCircle(this.geometry.coordinates, wp.geometry.coordinates, {
        npoints: 8,
        properties
      })
    );

    gc.properties = {
      ...gc.properties,
      distance: this.distanceTo(wp),
      bearing: this.bearingTo(wp)
    };

    return gc;
  }

  destinationPoint(distance, heading, properties) {
    const coords = rhumbDestination(
      this.geometry.coordinates,
      new Distance(distance),
      new Azimuth(heading).to("°"),
      {
        units: "meters"
      }
    ).geometry.coordinates;
    if (this.geometry.coordinates[2]) coords[2] = this.geometry.coordinates[2];
    return new this.constructor(coords, properties);
  }

  toGeoJSON(type) {
    switch (type) {
      case "Position":
        return this.geometry.coordinates;
      case "Point":
        return this.geometry;
      default:
        return {
          ...this,
          type: "Feature"
        };
    }
  }

  static from({
    type,
    geometry,
    coordinates,
    properties = {},
    ...otherProps
  } = {}) {
    if (arguments[0] instanceof this) return arguments[0];

    switch (type) {
      case this.name:
      case "Feature":
        if (geometry && geometry.type == "Point")
          coordinates = geometry.coordinates;
        else
          throw `Invalid data : 'geometry.type' should be 'Point' got '${geometry &&
          geometry.type}'`;
        break;
      case "Point":
        // coordinates = coordinates;
        break;
      default:
        throw `Invalid data : 'type' should be 'Feature', 'Point' or '${this.name}' got '${type}'`;
    }
    return new this(coordinates, {
      ...otherProps,
      ...properties
    });
  }

  static fromEvent({ lngLat: { lng, lat } } = {}) {
    return new this([lng, lat]);
  }
}
