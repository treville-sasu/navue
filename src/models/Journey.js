import { Model } from "@/models/Base";
import { Branch } from "@/models/Branch";
import { Distance } from "@/models/Quantities";

import { multiLineString, featureCollection, round } from "@turf/helpers";
import bbox from "@turf/bbox";
import length from "@turf/length";

export { Branch };
// FIXME: create branche if none exists
export class Journey extends Model {
  constructor(properties = {}, ...branches) {
    super({ properties, branches });
  }

  get name() {
    return this.properties.name;
  }

  set name(val) {
    return (this.properties.name = val);
  }

  last() {
    return this.branches[this.branches.length - 1];
  }

  push(...positions) {
    return this.last().push(...positions);
  }

  shift() {
    return this.last().shift();
  }

  forEach(cb) {
    return this.branches.forEach(cb);
  }

  map(cb) {
    return this.branches.map(cb);
  }

  filter(cb) {
    return this.branches.filter(cb);
  }

  addBranch(locations = [], properties) {
    this.branches.push(new Branch(undefined, properties, ...locations));
    return this;
  }

  getBranch(ref) {
    if (ref instanceof Branch) return ref;
    else if (Number.isInteger(ref)) return this.branches[ref];
    else return undefined;
  }

  removeBranch(ref) {
    let index = this.branches.indexOf(this.getBranch(ref));
    this.branches.splice(index, 1);
    return this;
  }

  get geom() {
    return this.map(b => b.geom);
  }

  get bbox() {
    return bbox(multiLineString(this.geom));
  }

  get length() {
    return new Distance(
      round(length(multiLineString(this.geom), { units: "meters" }))
    );
  }

  //TODO Specs multi types
  toGeoJSON(...types) {
    if (types.length > 1) {
      return featureCollection(
        types.flatMap(type => {
          let feat = this.toGeoJSON(type);
          return feat.features ? feat.features : feat;
        })
      );
    } else
      switch (types[0]) {
        case "Feature":
        case "LineString":
          return featureCollection(
            this.branches.flatMap(b => b.toGeoJSON("LineString"), {
              bbox: this.bbox
            })
          );
        case "MultiLineString":
          return featureCollection(
            this.branches.flatMap(
              b => b.toGeoJSON("MultiLineString").features,
              {
                bbox: this.bbox
              }
            )
          );
        case "FeatureCollection":
        default:
          return featureCollection(
            this.branches.flatMap(
              b => b.toGeoJSON("FeatureCollection").features
            ),
            { bbox: this.bbox }
          );
      }
  }

  static from({ type, branches = [], properties, ...others }) {
    if (arguments[0] instanceof this) return arguments[0];
    else if (type === this.name) {
      let created = new this(properties);
      branches.forEach(({ features, properties }) => {
        created.addBranch(features, properties);
      });

      Object.assign(created, others); //allow for private props like _id, _rev....
      return created;
    } else throw `Invalid data : 'type' should be '${this.name}' got '${type}'`;
  }
}
