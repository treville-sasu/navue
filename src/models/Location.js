import { Waypoint } from "@/models/Waypoint.js";
import { Altitude, Distance, Speed, Azimuth } from "@/models/Quantities.js";

import bbox from "@turf/bbox";
import buffer from "@turf/buffer";

export class Location extends Waypoint {
  get bbox() {
    return bbox(
      buffer(this.toGeoJSON(), Number(this.properties.accuracy), {
        units: "meters"
      })
    );
  }

  bounds(radius) {
    return bbox(
      buffer(this.toGeoJSON(), radius, {
        units: "meters"
      })
    );
  }

  legFrom(other) {
    return {
      distance: other.distanceTo(this),
      heading: other.bearingTo(this),
      speed: this._speedFrom(other),
      verticalSpeed: this._verticalSpeedFrom(other)
    };
  }

  willBeIn(delay) {
    if (this.properties.speed && this.properties.heading)
      return this.destinationPoint(
        this.properties.speed * delay,
        this.properties.heading
      );
    else
      throw "cannot calculate futur position, speed & heading should be provided";
  }

  _delayFrom({ properties: { timestamp } }) {
    return (this.properties.timestamp - timestamp) / 1000;
  }

  _climbFrom({ properties: { altitude } }) {
    return this.properties.altitude - altitude;
  }

  _speedFrom(other) {
    if (this._delayFrom(other) == 0)
      throw "Teleportation is not yet a possibility";
    return new Speed(this.distanceTo(other) / this._delayFrom(other));
  }

  _verticalSpeedFrom(other) {
    if (this._delayFrom(other) == 0)
      throw "Teleportation is not yet a possibility";

    return new Speed(this._climbFrom(other) / this._delayFrom(other));
  }

  static from({ accuracy, altitude, speed, heading, ...others } = {}) {
    if (arguments[0] instanceof this) return arguments[0];
    if (accuracy) accuracy = Distance.from(accuracy);
    if (altitude) altitude = Altitude.from(altitude);
    if (speed) speed = Speed.from(speed);
    if (heading) heading = Azimuth.from(heading);
    return super.from({ accuracy, altitude, speed, heading, ...others });
  }

  static fromGeolocationPosition({
    coords: {
      latitude,
      longitude,
      altitude,
      accuracy,
      altitudeAccuracy,
      heading,
      speed
    },
    timestamp
  }) {
    if (altitude)
      altitude = new Altitude(altitude, "m", {
        reference: "WGS84",
        accuracy: altitudeAccuracy
      });

    if (accuracy) accuracy = new Distance(accuracy, "m");
    if (speed) speed = new Speed(speed, "m/s");
    if (heading) heading = new Azimuth(heading, "Â°");

    return new this([longitude, latitude, altitude], {
      accuracy,
      altitude,
      speed,
      heading,
      timestamp
    });
  }
}
