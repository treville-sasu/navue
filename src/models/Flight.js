import { Journey } from "@/models/Journey";
import { Location } from "@/models/Location";
import { Branch } from "./Journey";

export class Flight extends Journey {
  constructor(
    { name, notes, checked = [], milestones = [] } = {},
    ...locations
  ) {
    // let branches;
    // if (locations.every(a => a instanceof Location))
    //   branches = [new Branch(Location, {}, ...locations)];
    // else if (
    //   locations.every(
    //     a => a instanceof Array && a.every(b => b instanceof Branch)
    //   )
    // )
    //   branches = locations.flat();
    let branches = locations;
    super({ name, notes, checked, milestones }, ...branches);
  }

  addBranch(locations, properties) {
    this.branches.push(new Branch(Location, properties, ...locations));
    return this;
  }

  get duration() {
    return (
      this.branches.reduce((total, b) => {
        return (total +=
          b.last().properties.timestamp - b.first().properties.timestamp);
      }, 0) / 1000
    );
  }

  // static from({ type, locations, features, ...properties } = {}) {
  //   if (arguments[0] instanceof this) return arguments[0];

  //   switch (type) {
  //     case this.name:
  //       if (locations)
  //         locations = locations.map(t => t.map(l => Location.from(l)));

  //       break;
  //     case "FeatureCollection":
  //       locations = [features.map(l => Location.from(l))];
  //       // properties = { ...properties };
  //       break;

  //     default:
  //       throw `Invalid data : 'type' should be 'FeatureCollection' or '${this.name}' got '${type}'`;
  //   }

  //   return new this(properties, ...locations);
  // }
}
