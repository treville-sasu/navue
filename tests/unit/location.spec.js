/* eslint-disable jest/no-hooks */
import { Waypoint } from "@/models/Waypoint";
import { Location } from "@/models/Location";
import { Distance, Speed, Azimuth } from "@/models/Quantities";

describe("location", () => {
  let location;

  beforeEach(() => {
    location = new Location([0, 0, 0], {
      accuracy: new Distance(20),
      timestamp: 1234567890,
      name: "Echo",
      notes: "Some note"
    });
  });

  it("extends Waypoint", () => {
    expect(Location.prototype).toBeInstanceOf(Waypoint);
  });

  it("is instantiated", () => {
    expect(() => new Location()).toThrowError();
    expect(() => new Location("aaaa")).toThrowError();
    expect(location).toMatchSnapshot();
  });

  it("return a bounding box from accuracy", () => {
    expect(location.bbox).toMatchSnapshot();
  });

  it("return a bounding box from radius", () => {
    expect(location.bounds(200)).toMatchSnapshot();
  });

  it("return a futur location in time (seconds later)", () => {
    expect(() => {
      location.willBeIn(60);
    }).toThrow(
      "cannot calculate futur position, speed & heading should be provided"
    );

    location.properties.speed = new Speed(10);
    expect(() => {
      location.willBeIn(60);
    }).toThrow(
      "cannot calculate futur position, speed & heading should be provided"
    );

    location.properties.heading = new Azimuth(90);
    expect(location.willBeIn(60)).toMatchSnapshot();
  });

  it("retun a leg from other position", () => {
    const prevLocation = new Location([1, 1, 1], {
      accuracy: new Distance(20),
      timestamp: 1234567890
    });

    expect(() => location.legFrom(prevLocation)).toThrow(
      "Teleportation is not yet a possibility"
    );

    prevLocation.properties.timestamp -= 10000;

    expect(location.legFrom(prevLocation)).toMatchInlineSnapshot(`
      Object {
        "distance": Object {
          "precision": 5,
          "type": "Distance",
          "unit": "m",
          "value": 157250,
        },
        "heading": Object {
          "precision": 16,
          "type": "Azimuth",
          "unit": "°",
          "value": 224.9985454851105,
        },
        "speed": Object {
          "precision": 5,
          "type": "Speed",
          "unit": "m/s",
          "value": 15725,
        },
        "verticalSpeed": Object {
          "precision": 1,
          "type": "Speed",
          "unit": "m/s",
          "value": -0.1,
        },
      }
    `);
  });

  it("imports from literal", () => {
    expect(
      Location.from({
        longitude: 0,
        latitude: 43,
        altitude: {
          accuracy: 50,
          reference: "WGS84",
          value: 1000,
          unit: "m",
          type: "Altitude"
        },
        accuracy: {
          value: 100,
          unit: "m",
          type: "Distance"
        },
        speed: {
          value: 10,
          unit: "m/s",
          type: "Speed"
        },
        heading: {
          value: 180,
          unit: "°",
          type: "Azimuth"
        },
        timestamp: 1620672163861,
        type: "Location",
        _id: "1620672163861",
        _rev: "1-ad4a6ac12faa5c391774a3120af8c39d"
      })
    ).toMatchSnapshot();
  });

  it("imports from GeolocationPosition", () => {
    expect(
      Location.fromGeolocationPosition({
        coords: {
          latitude: 0,
          longitude: 0,
          altitude: 10,
          accuracy: 100,
          altitudeAccuracy: 10,
          heading: 270,
          speed: 10
        },
        timestamp: 1234567890
      })
    ).toMatchSnapshot();
  });
});
