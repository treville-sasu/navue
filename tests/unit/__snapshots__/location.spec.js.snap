// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`location imports from GeolocationPosition 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      0,
      0,
      10,
    ],
    "type": "Point",
  },
  "properties": Object {
    "accuracy": Object {
      "precision": 1,
      "type": "Distance",
      "unit": "m",
      "value": 100,
    },
    "altitude": Object {
      "accuracy": 10,
      "precision": 1,
      "reference": "WGS84",
      "type": "Altitude",
      "unit": "m",
      "value": 10,
    },
    "heading": Object {
      "precision": 2,
      "type": "Azimuth",
      "unit": "°",
      "value": 270,
    },
    "speed": Object {
      "precision": 1,
      "type": "Speed",
      "unit": "m/s",
      "value": 10,
    },
    "timestamp": 1234567890,
  },
  "type": "Location",
}
`;

exports[`location imports from literal 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      0,
      43,
      1000,
    ],
    "type": "Point",
  },
  "properties": Object {
    "_id": "1620672163861",
    "_rev": "1-ad4a6ac12faa5c391774a3120af8c39d",
    "accuracy": Object {
      "precision": 1,
      "type": "Distance",
      "unit": "m",
      "value": 100,
    },
    "altitude": Object {
      "accuracy": 50,
      "precision": 1,
      "reference": "WGS84",
      "type": "Altitude",
      "unit": "m",
      "value": 1000,
    },
    "heading": Object {
      "precision": 2,
      "type": "Azimuth",
      "unit": "°",
      "value": 180,
    },
    "speed": Object {
      "precision": 1,
      "type": "Speed",
      "unit": "m/s",
      "value": 10,
    },
    "timestamp": 1620672163861,
  },
  "type": "Location",
}
`;

exports[`location is instantiated 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      0,
      0,
      0,
    ],
    "type": "Point",
  },
  "properties": Object {
    "accuracy": Object {
      "precision": 1,
      "type": "Distance",
      "unit": "m",
      "value": 20,
    },
    "altitude": Object {
      "precision": 0,
      "reference": undefined,
      "type": "Altitude",
      "unit": "m",
      "value": 0,
    },
    "name": "Echo",
    "notes": "Some note",
    "timestamp": 1234567890,
  },
  "type": "Location",
}
`;

exports[`location return a bounding box from accuracy 1`] = `
Array [
  -0.0001798640727449076,
  -0.0001798640727449076,
  0.0001798640727449076,
  0.0001798640727449076,
]
`;

exports[`location return a bounding box from radius 1`] = `
Array [
  -0.001798640727449076,
  -0.001798640727449076,
  0.001798640727449076,
  0.001798640727449076,
]
`;

exports[`location return a futur location in time (seconds later) 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      0.005396,
      0,
    ],
    "type": "Point",
  },
  "properties": Object {},
  "type": "Location",
}
`;
